File: 0047-permutations-ii/0047-permutations-ii.rb
# @param {Integer[]} nums
# @return {Integer[][]}
def permute_unique(nums)
    return [nums] if nums.length == 1
    
    uniq_perms, dupe, ans = Set.new, [], []
    nums.each_with_index do |num, i|
        copy = nums.map(&:clone)
        ext = copy.delete_at(i)
        permute_unique(copy).each{ |p| dupe << [ext] + p }
    end
    
    dupe.each do |d|
        if !uniq_perms.include?(d)
            ans << d
            uniq_perms.add(d)
        end
    end
    
    return ans
end

File: 0047-permutations-ii/0047-permutations-ii.py
class Solution:
    def permuteUnique(self, nums: List[int]) -> List[List[int]]:
        if len(nums) == 1:
            return [nums]
        
        uniqPerms, dupe, ans = set(), [], []
        for i, num in enumerate(nums):
            lCopy = nums.copy()
            ext = lCopy.pop(i)
            for p in self.permuteUnique(lCopy):
                dupe.append([ext] + p)
                
        for d in dupe:
            if tuple(d) not in uniqPerms:
                ans.append(d)
                uniqPerms.add(tuple(d))
        
        return ans
            